README
NAME
    Perl::Critic::Policy::logicLAB::RequireSheBang - simple policy for
    keeping your shebang line uniform

AFFILIATION
    This policy is a policy in the Perl::Critic::logicLAB distribution. The
    policy is themed: logiclab.

VERSION
    This documentation describes version 0.05.

DESCRIPTION
    This policy is intended in guarding your use of the shebang line. It
    assists in making sure that your shebang line adheres to certain
    formats.

    The default format is

        #!/usr/local/bin/perl

    You can however specify another or define your own in the configuration
    of the policy.

    NB this policy does currently not warn about missing shebang lines, it
    only checks shebang lines encountered.

CONFIGURATION AND ENVIRONMENT
    This policy allow you to configure the contents of the shebang lines you
    want to allow using "formats".

  formats
        [logicLAB::RequireSheBang]
        formats = #!/usr/local/bin/perl || #!/usr/bin/perl || #!perl || #!env perl

    Since the default shebang line enforced by the policy is:

        #!/usr/local/bin/perl

    Please note that if you however what to extend the pattern, you also
    have to specify was is normally the default pattern since configuration
    overwrites the default even for extensions.

    This mean that if you want to allow:

        #!/usr/local/bin/perl

        #!/usr/local/bin/perl -w

        #!/usr/local/bin/perl -wT

    Your format should look like the following:

        [logicLAB::RequireSheBang]
        formats = #!/usr/local/bin/perl || #!/usr/local/bin/perl -w || #!/usr/local/bin/perl -wT

  exempt_modules
    You can specify if you want to check modules also. The default is to
    exempt from checking shebang lines in modules.

            [logicLAB::RequireSheBang]
            exempt_modules = 0

  debug
    Optionally and for development purposes I have added a debug flag. This
    can be set in your Perl::Critic configuration file as follows:

        [logicLAB::RequireSheBang]
        debug = 1

    This enables more explicit output on what is going on during the actual
    processing of the policy.

DEPENDENCIES AND REQUIREMENTS
    *   Perl::Critic

    *   Perl::Critic::Utils

    *   Readonly

    *   Test::More

    *   Test::Perl::Critic

    *   List::MoreUtils

INCOMPATIBILITIES
    This distribution has no known incompatibilities.

BUGS AND LIMITATIONS
    The distribution has now known bugs or limitations. It locates shebang
    lines through out the source code, not limiting itself to the first
    line. This might however change in the future, but will propably be made
    configurable if possible.

BUG REPORTING
    Please use Requets Tracker for bug reporting:

    *   <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Perl-Critic-logicLAB-Requi
        reSheBang>

TEST AND QUALITY
    The following policies have been disabled for this distribution

    *   Perl::Critic::Policy::ValuesAndExpressions::ProhibitConstantPragma

        Constants are good, - see the link below.

        *   <https://logiclab.jira.com/wiki/display/OPEN/Perl-Critic-Policy-
            ValuesAndExpressions-ProhibitConstantPragma>

    *   Perl::Critic::Policy::NamingConventions::Capitalization

    *   Data::Dumper

    See also t/perlcriticrc

  TEST COVERAGE
    Coverage test executed the following way, the coverage report is based
    on the version described in this documentation (see "VERSION").

        ./Build testcover

            ---------------------------- ------ ------ ------ ------ ------ ------ ------
            File                           stmt   bran   cond    sub    pod   time  total
            ---------------------------- ------ ------ ------ ------ ------ ------ ------
            ...ogicLAB/RequireSheBang.pm   70.4   64.3   44.4  100.0  100.0  100.0   72.1
            Total                          70.4   64.3   44.4  100.0  100.0  100.0   72.1
            ---------------------------- ------ ------ ------ ------ ------ ------ ------

SEE ALSO
    *   Perl::Critic

    *   <http://perldoc.perl.org/perlrun.html>

    *   <http://logiclab.jira.com/wiki/display/OPEN/Development#Development-
        MakeyourComponentsEnvironmentAgnostic>

    *   <http://logiclab.jira.com/wiki/display/PCPLRSB/Home>

    *   <http://logiclab.jira.com/wiki/display/PCLL/Home>

AUTHOR
    *   Jonas B. Nielsen, jonasbn "<jonasbn@cpan.org>"

ACKNOWLEDGEMENT
    *   Erik Johansen (uniejo), feedback to version 0.01

LICENSE AND COPYRIGHT
    Copyright (c) 2011-2014 Jonas B. Nielsen, jonasbn. All rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Changes file for Perl::Critic::Policy::logicLAB::RequireSheBang

Project change log:

http://logiclab.jira.com/browse/PCPLRSB#selectedTab=com.atlassian.jira.plugin.system.project%3Achangelog-panel

0.06 2014.08.11

- Maintenance release, update not necessary

- Removed dot dirs from distribution

0.05 2014-02-27

- Maintenance release, update not necessary

- Specified Module::Build 0.40 requirement as hinted by Module::Build

- Specified Perl version 5.8.0 in Build.PL

- Added new feature: exemption from package as a configurable parameter
  This feature makes it possible to avoid the check from package files
  [PCPLRSB-5]

  https://logiclab.jira.com/browse/PCPLRSB-5

- Added kwalitee.t implementing Test::Kwalitee tests [PCPLRSB-12]

  https://logiclab.jira.com/browse/PCPLRSB-12

- Added changes.t automatic assertion of the integrity of the Changes 
  file, see also PCPLRSB-10 [PCPLRSB-11]

  https://logiclab.jira.com/browse/PCPLRSB-11


0.04 2013-08-02

- Maintenance release, update not necessary

- Adding a Changes file as part of my Questhub quest adhering to the standard
  described in: CPAN::Changes::Spec [PCPLRSB-10]

  Quest: http://questhub.io/realm/perl/quest/51f0fd8e18ba7dcf4b000040
  Spec: https://metacpan.org/module/CPAN::Changes::Spec

  https://logiclab.jira.com/browse/PCPLRSB-10

- Minor improvement/correction to the central regular expression [PCPLRSB-9]

  https://logiclab.jira.com/browse/PCPLRSB-9

- Added a debugging flag [PCPLRSB-8]

  https://logiclab.jira.com/browse/PCPLRSB-8


0.03 2011-06-06

- Bug fix release, update recommended

- Serious bug fix to parsing of shebang lines [PCPLRSB-6]
  http://logiclab.jira.com/browse/PCPLRSB-6


0.02 2011-05-16

- Bug fix release, Update recommended

- Minor corrections to documentation [PCPLRSB-3]
  http://logiclab.jira.com/browse/PCPLRSB-3

  Theme changed to logiclab from maintenance according to the guidelines, I
  had previously overlooked this:
  http://search.cpan.org/dist/Perl-Critic/lib/Perl/Critic/DEVELOPER.pod#Themes

- Added Changes (this file)


0.01 2011-05-03

- Initial release


Build.PL
# Build.PL for Perl::Critic::Policy::logicLAB::RequireSheBang

# $Id$

use strict;
use Module::Build;

my $build = Module::Build->new(
    dist_author          => 'Jonas B. Nielsen, <jonasbn@cpan.org>',
    module_name          => 'Perl::Critic::Policy::logicLAB::RequireSheBang',
    license              => 'artistic',
    recursive_test_files => 0,
    create_makefile_pl   => 'traditional',
    create_readme        => 1,
    dist_abstract        => q[simple policy asserting she bang line],
    configure_requires => { 'Module::Build' => 0.40 },
    requires => {
        'Perl::Critic'        => '0',
        'Perl::Critic::Utils' => '0',
        'Readonly'            => '0',
        'Test::More'          => '0',
        'Test::Perl::Critic'  => '0',
        'List::MoreUtils'     => '0',
        'perl'                => '5.8.0',
    },
    meta_merge => {
        resources => {
            homepage => 'http://logiclab.jira.com/wiki/display/PCPLRSB/Home',
            bugtracker => 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Perl-Critic-Policy-logicLAB-Require-SheBang',
            repository => 'https://github.com/jonasbn/pcplrsb',
            ProjectChangelog =>
                'http://logiclab.jira.com/browse/PCPLRSB#selectedTab=com.atlassian.jira.plugin.system.project%3Achangelog-panel',
            ProjectSite => 'http://logicLAB.jira.com/browse/PCPLRSB',
            ProjectWiki => 'http://logiclab.jira.com/wiki/display/PCPLRSB/Home',
        }
    },
);

$build->create_build_script();

1;


lib/Perl/Critic/Policy/logicLAB/RequireSheBang.pm
package Perl::Critic::Policy::logicLAB::RequireSheBang;

# $Id$

use strict;
use warnings;
use base 'Perl::Critic::Policy';
use Perl::Critic::Utils qw{ $SEVERITY_MEDIUM :booleans };
use List::MoreUtils qw(none);
use Data::Dumper;
use 5.008;

$Data::Dumper::Useqq = 1;

our $VERSION = '0.06';

Readonly::Scalar my $EXPL  => q{she-bang line should adhere to requirement};
Readonly::Scalar my $DEBUG => q{DEBUG logicLAB::RequireSheBang};

use constant default_severity     => $SEVERITY_MEDIUM;
use constant default_themes       => qw(logiclab);
use constant supported_parameters => qw(formats debug);

sub prepare_to_scan_document {
    my ( $self, $document ) = @_;
 	if ($self->{exempt_modules} && $document->is_module()) {
		return 0;
	}

    return $document->is_program();
}

sub violates {
    my ( $self, $element, $doc ) = @_;

	my $statement = $doc->find_first( 'PPI::Token::Comment' );

	if (not $statement->location()->[0]) {
        return $self->violation(
            q{she-bang line not located as first line},
            $EXPL, $statement );
	}

    if ( $self->{debug} ) {
        print {*STDERR} "$DEBUG: we got statement:\n";
        print {*STDERR} Dumper $statement;
    }

    my ( $shebang, $cli ) = $element =~ m{
            \A  #beginning of string
            (\#!) #actual she-bang
            #([^\r\n]*?) #the path and possible flags
            ([/\-\w ]+?) #the path and possible flags, note the space character
            \s* #possible left over whitespace (PPI?)
            \Z #indication of end of string to assist above capture
    }xsm;

    if ($cli) {
        $cli =~ s{
            \s+ #one or more whitespace character, PCPLRSB-9 / http://logiclab.jira.com/browse/PCPLRSB-9
            $ #end of string
        }{}xsm;
    }

    if ( $self->{debug} && $shebang && $cli ) {
        print {*STDERR} "$DEBUG: we got a shebang line:\n";
        print {*STDERR} '>' . $shebang . $cli . "<\n";

        print {*STDERR} "$DEBUG: comparing against formats:\n";
        print {*STDERR} Dumper $self->{_formats};
        print {*STDERR} "\n";

    } elsif ( $self->{debug} ) {
        print {*STDERR} "$DEBUG: not a shebang, ignoring...\n";
    }

    if ( $shebang && none { ( $shebang . $cli ) eq $_ }
        @{ $self->{_formats} } )
    {

        if ( $self->{debug} ) {
            print {*STDERR} "$DEBUG: we got a violation:\n";
            print {*STDERR} '>' . $shebang . $cli . "<\n";
        }

        return $self->violation(
            q{she-bang line not conforming with requirement},
            $EXPL, $element );
    }

    return;
}

sub initialize_if_enabled {
    my ( $self, $config ) = @_;

    #Formats:
    #Setting the default
    $self->{_formats} = [ ('#!/usr/local/bin/perl') ];

    #fetching configured formats
    my $formats = $config->get('formats');

    #parsing configured formats, see also _parse_formats
    if ($formats) {
        $self->{_formats} = $self->_parse_formats($formats);
    }

    #debug
    $self->{debug} = $config->get('debug') || 0;

    #exempt_modules
    $self->{exempt_modules} = $config->get('exempt_modules') || 1;


    return $TRUE;
}

sub _parse_formats {
    my ( $self, $config_string ) = @_;

    my @formats = split m{ \s* [||]+ \s* }xsm, $config_string;

    return \@formats;
}

1;

__END__

=pod

=head1 NAME

Perl::Critic::Policy::logicLAB::RequireSheBang - simple policy for keeping your shebang line uniform

=head1 AFFILIATION

This policy is a policy in the Perl::Critic::logicLAB distribution. The policy
is themed: logiclab.

=head1 VERSION

This documentation describes version 0.05.

=head1 DESCRIPTION

This policy is intended in guarding your use of the shebang line. It assists
in making sure that your shebang line adheres to certain formats.

The default format is

    #!/usr/local/bin/perl

You can however specify another or define your own in the configuration of the
policy.

B<NB> this policy does currently not warn about missing shebang lines, it only
checks shebang lines encountered.

=head1 CONFIGURATION AND ENVIRONMENT

This policy allow you to configure the contents of the shebang lines you
want to allow using L</formats>.

=head2 formats

    [logicLAB::RequireSheBang]
    formats = #!/usr/local/bin/perl || #!/usr/bin/perl || #!perl || #!env perl

Since the default shebang line enforced by the policy is:

    #!/usr/local/bin/perl

Please note that if you however what to extend the pattern, you also have
to specify was is normally the default pattern since configuration
overwrites the default even for extensions.

This mean that if you want to allow:

    #!/usr/local/bin/perl

    #!/usr/local/bin/perl -w

    #!/usr/local/bin/perl -wT

Your format should look like the following:

    [logicLAB::RequireSheBang]
    formats = #!/usr/local/bin/perl || #!/usr/local/bin/perl -w || #!/usr/local/bin/perl -wT

=head2 exempt_modules

You can specify if you want to check modules also. The default is to exempt from checking
shebang lines in modules.

	[logicLAB::RequireSheBang]
	exempt_modules = 0

=head2 debug

Optionally and for development purposes I have added a debug flag. This can be set in
your L<Perl::Critic> configuration file as follows:

    [logicLAB::RequireSheBang]
    debug = 1

This enables more explicit output on what is going on during the actual processing of
the policy.

=head1 DEPENDENCIES AND REQUIREMENTS

=over

=item * L<Perl::Critic>

=item * L<Perl::Critic::Utils>

=item * L<Readonly>

=item * L<Test::More>

=item * L<Test::Perl::Critic>

=item * L<List::MoreUtils>

=back

=head1 INCOMPATIBILITIES

This distribution has no known incompatibilities.

=head1 BUGS AND LIMITATIONS

The distribution has now known bugs or limitations. It locates shebang lines
through out the source code, not limiting itself to the first line. This might
however change in the future, but will propably be made configurable if possible.

=head1 BUG REPORTING

Please use Requets Tracker for bug reporting:

=over

=item * L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Perl-Critic-logicLAB-RequireSheBang>

=back

=head1 TEST AND QUALITY

The following policies have been disabled for this distribution

=over

=item * L<Perl::Critic::Policy::ValuesAndExpressions::ProhibitConstantPragma>

Constants are good, - see the link below.

=over

=item * L<https://logiclab.jira.com/wiki/display/OPEN/Perl-Critic-Policy-ValuesAndExpressions-ProhibitConstantPragma>

=back

=item * L<Perl::Critic::Policy::NamingConventions::Capitalization>

=item * L<Data::Dumper>

=back

See also F<t/perlcriticrc>

=head2 TEST COVERAGE

Coverage test executed the following way, the coverage report is based on the
version described in this documentation (see L</VERSION>).

    ./Build testcover

	---------------------------- ------ ------ ------ ------ ------ ------ ------
	File                           stmt   bran   cond    sub    pod   time  total
	---------------------------- ------ ------ ------ ------ ------ ------ ------
	...ogicLAB/RequireSheBang.pm   70.4   64.3   44.4  100.0  100.0  100.0   72.1
	Total                          70.4   64.3   44.4  100.0  100.0  100.0   72.1
	---------------------------- ------ ------ ------ ------ ------ ------ ------

=head1 SEE ALSO

=over

=item * L<Perl::Critic>

=item * L<http://perldoc.perl.org/perlrun.html>

=item * L<http://logiclab.jira.com/wiki/display/OPEN/Development#Development-MakeyourComponentsEnvironmentAgnostic>

=item * L<http://logiclab.jira.com/wiki/display/PCPLRSB/Home>

=item * L<http://logiclab.jira.com/wiki/display/PCLL/Home>

=back

=head1 AUTHOR

=over

=item * Jonas B. Nielsen, jonasbn C<< <jonasbn@cpan.org> >>

=back

=head1 ACKNOWLEDGEMENT

=over

=item * Erik Johansen (uniejo), feedback to version 0.01

=back

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2011-2014 Jonas B. Nielsen, jonasbn. All rights reserved.

This program is free software; you can redistribute it and/or modify it under the
same terms as Perl itself.

=cut
